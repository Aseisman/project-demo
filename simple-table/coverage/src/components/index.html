
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/components</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/components</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">99.51% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>617/620</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">96.55% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>56/58</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">86.95% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>20/23</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">99.51% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>617/620</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Body.tsx"><a href="Body.tsx.html">Body.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="37" class="abs high">37/37</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="37" class="abs high">37/37</td>
	</tr>

<tr>
	<td class="file high" data-value="Headers.tsx"><a href="Headers.tsx.html">Headers.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="95" class="abs high">95/95</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="19" class="abs high">19/19</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="95" class="abs high">95/95</td>
	</tr>

<tr>
	<td class="file high" data-value="Line.tsx"><a href="Line.tsx.html">Line.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="53" class="abs high">53/53</td>
	<td data-value="88.88" class="pct high">88.88%</td>
	<td data-value="9" class="abs high">8/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="53" class="abs high">53/53</td>
	</tr>

<tr>
	<td class="file high" data-value="SimpleTable.tsx"><a href="SimpleTable.tsx.html">SimpleTable.tsx</a></td>
	<td data-value="99.71" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 99%"></div><div class="cover-empty" style="width: 1%"></div></div>
	</td>
	<td data-value="99.71" class="pct high">99.71%</td>
	<td data-value="349" class="abs high">348/349</td>
	<td data-value="95.23" class="pct high">95.23%</td>
	<td data-value="21" class="abs high">20/21</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="99.71" class="pct high">99.71%</td>
	<td data-value="349" class="abs high">348/349</td>
	</tr>

<tr>
	<td class="file high" data-value="types.ts"><a href="types.ts.html">types.ts</a></td>
	<td data-value="97.67" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 97%"></div><div class="cover-empty" style="width: 3%"></div></div>
	</td>
	<td data-value="97.67" class="pct high">97.67%</td>
	<td data-value="86" class="abs high">84/86</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="97.67" class="pct high">97.67%</td>
	<td data-value="86" class="abs high">84/86</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at Mon May 30 2022 15:08:04 GMT+0800 (中国标准时间)
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    