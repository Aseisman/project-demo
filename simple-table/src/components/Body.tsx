import { defineComponent, onMounted, reactive, ref,PropType} from "vue";
import { type TableProps, type ColumnsProps, tableProps } from "./types";
import TLine from "./Line";

type BodyProps = {
  columns: ColumnsProps[];
  data: any[];
};

export default defineComponent({
  name: "Body",
  props: {
    columns: {
      type: Array as PropType<ColumnsProps[]>,
    },
    data: {
      type: Array as PropType<any[]>,
    },
  },
  //@ts-ignore
  setup(props: BodyProps) {
    const columns = () => {
      if (props.data?.length > 0) {
        return props.data.map((line: any, index: string | number) => {
          return <TLine line={line} columns={props.columns}></TLine>;
        });
      } else {
        return (
          <td style="min-height:300px" colspan={props.columns.length}>
            {"没有数据哦"}
          </td>
        );
      }
    };
    return () => columns();
  },
});