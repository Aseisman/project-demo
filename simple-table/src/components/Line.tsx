import { defineComponent, onMounted, reactive, ref ,PropType} from "vue";
import { type TableProps, tableProps } from "./types";
import { type ColumnsProps } from "./types";

type lineProps = {
  line: any;
  columns: ColumnsProps[];
  index?: number|string;
};

export default defineComponent({
  name: "Line",
  props: {
    line: {
      type: Object as PropType<any>,
      // default: () => ({}),
    },
    columns: {
      type: Array as PropType<ColumnsProps[]>,
      // default: () => [],
    },
    index: {
      type: [Number, String],
    },
  },
  //@ts-ignore
  setup(props: lineProps) {
    const columns = () => {
      let tds = props.columns.map((column: ColumnsProps) => {
        let field: string = column.field;
        let formatter: Function| null = column.formatter;
        let render: Function | null = column.render;
        if (render) {
          return <td>{render(props.line[field], props.line, props.index)}</td>;
        }
        return (
          <td>
            <div>
              <div>
                {formatter ? formatter(props.line[field]) : props.line[field]||''}
              </div>
            </div>
          </td>
        );
      });
      return <tr>{tds}</tr>;
      // return <tr>{
      //   [1,2,3,4].map(item=><td>{item}</td>)
      //   }</tr>
    };
    return () => columns();
  },
});
