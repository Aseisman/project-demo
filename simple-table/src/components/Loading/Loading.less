.loading{
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 1004;
    display: flex;
    align-items: center;
    justify-content: center;
    background: #9d9f9d38;
}