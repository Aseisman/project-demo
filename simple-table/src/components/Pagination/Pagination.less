.pagination-box {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  margin-top: 8px;
  .pagination-number {
    margin-right: 8px;
    width: 30px;
    height: 32px;
    border: 1px solid #d9d9d9;
    line-height: 32px;
    color: rgb(53, 53, 53);
    border-radius: 4px;
    // color: #1890ff;
  }
  .pagination-number-active {
    margin-right: 8px;
    width: 30px;
    height: 32px;
    border: 1px solid #1890ff;
    line-height: 32px;
    color: #1890ff;
    border-radius: 4px;
    cursor: pointer;
  }
  .pagination-number:hover {
    cursor: pointer;
    color: #1890ff;
    border: 1px solid #1890ff;
  }
  .pagination-pre {
    margin-right: 8px;
    cursor: pointer;
  }
  .pagination-pre-disabled {
    margin-right: 8px;
    cursor: not-allowed;
    color: #d9d9d9;
  }
  .pagination-aft-disabled {
    margin-right: 8px;
    cursor: not-allowed;
    color: #d9d9d9;
  }
  .pagination-aft {
    margin-right: 8px;
    cursor: pointer;
  }
}
