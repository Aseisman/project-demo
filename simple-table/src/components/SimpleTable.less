table {
  // border: 1px solid #858587; /*设置table的边框*/
  border-collapse: collapse; /*边框合并*/
}

th,
td {
  border: 1px solid #858587; /*为单元格单独设置边框*/
  border-collapse: collapse; /*边框合并*/
  line-height: 30px;
}

.triangle-up-blue {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-bottom: 8px solid rgb(78, 155, 255);
  cursor: pointer;
}
.triangle-up {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-bottom: 8px solid #858587;
  cursor: pointer;
}
.triangle-down {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-top: 8px solid #858587;
  cursor: pointer;
}
.triangle-down-blue {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-top: 8px solid rgb(78, 155, 255);
  cursor: pointer;
}
