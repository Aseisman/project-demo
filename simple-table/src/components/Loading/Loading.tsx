import { defineComponent, onMounted, reactive, ref } from "vue";
import "./Loading.less";
type LoadProps = {
  loading: Boolean
}
export default defineComponent({
  name: "Loading",
  props: {
    loading: Boolean,
  },
  setup(props: LoadProps, { attrs, emit, slots }) {
    return () => {        
      return (
        <div style={"position:relative;min-height:300px;"}>
          {props.loading ? <div class="loading">加载中</div> : ""}
          <div>{slots.default ? slots.default() : ""}</div>
        </div>
      );
    };
  },
});
