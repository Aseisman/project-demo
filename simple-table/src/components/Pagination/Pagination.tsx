import { defineComponent, onMounted, PropType, reactive, Ref, ref, watch } from "vue";
import { PaginationProps } from "../types";
import { debounce } from "../../utils";
import "./Pagination.less";

type PagProps = {
  pagination: PaginationProps,
}

export default defineComponent({
  name: "Pagination",
  props: {
    pagination: {
      type: Object as PropType<PaginationProps>,
      // default: () => ({}),
    },
  },
  emits: ["change"],
  //@ts-ignore
  setup(props: PagProps, { attrs, emit, slots }) {
    const change = (index: number) => {
      emit("change", "pageChange", index);
    };
    const inputChange = debounce((v: HTMLInputElement | any, nums: number) => {
      let value = v?.target?.value;
      if(value ==''){
        return change(0);
      }      
      if (value <= nums && value > 0) {
        change(value - 1);
      }else{
        console.error('输入的页数有问题');
      }
    }, 800);
    const templ = () => {
      let current = props.pagination.current || 0;
      let total = props.pagination.total;
      let pageSize = props.pagination.pageSize;
      let nums = Math.ceil(total / pageSize);
      let tpl = [];
      tpl.push(<div style={"margin-right:8px;"}>共：{total}条数据</div>);
      tpl.push(
        <div
          class={current <= 0 ? "pagination-pre-disabled" : "pagination-pre"}
          onClick={() => {
            if (current > 0) change(current - 1);
          }}
          id='pre-btn'
        >
          后退
        </div>
      );
      for (let i = 1; i <= nums; i++) {
        tpl.push(
          <div
            class={
              current + 1 == i
                ? "pagination-number-active"
                : "pagination-number"
            }
            onClick={() => change(i - 1)}
          >
            {i}
          </div>
        );
      }
      tpl.push(
        <div
          class={
            current + 1 >= nums ? "pagination-aft-disabled" : "pagination-aft"
          }
          onClick={() => {
            if (current < nums - 1) change(current + 1);
          }}
          id='next-btn'
        >
          前进
        </div>
      );
      tpl.push(
        <div>
          跳至{" "}
          <input
            type="text"
            // @ts-ignore
            oninput={(v) => inputChange(v, nums)}
          />
          页
        </div>
      );
      return tpl;
    };
    return () => {
      return <div class="pagination-box">{templ()}</div>;
    };
  },
});
// export default defineComponent({
//     name: "Pagination",
//     props: {
//       pagination: {
//         type: [Object, Boolean],
//         default: () => ({}),
//       },
//     },
//     emits: ["change", "update:pagination"],
//     setup(props, { attrs, emit, slots }) {
//       const pagination = useVModel(props, "pagination", emit, {
//         passive: true,
//         deep: true,
//       });
//       const change = (index: number) => {
//         pagination.value.current = index;
//         emit("change", "pageChange", index);
//       };
//       const templ = () => {
//         let current = pagination.value.current || 0;
//         let total = pagination.value.total;
//         let pageSize = pagination.value.pageSize;
//         let nums = Math.ceil(total / pageSize);
//         let tpl = [];
//         tpl.push(<div style={"margin-right:8px;"}>共：{total}条数据</div>);
//         tpl.push(
//           <div
//             class={current <= 0 ? "pagination-pre-disabled" : "pagination-pre"}
//           >
//             后退
//           </div>
//         );
//         for (let i = 1; i <= nums; i++) {
//           tpl.push(
//             <div
//               class={
//                 current + 1 == i
//                   ? "pagination-number-active"
//                   : "pagination-number"
//               }
//               onClick={() => change(i - 1)}
//             >
//               {i}
//             </div>
//           );
//         }
//         tpl.push(
//           <div
//             class={
//               current + 1 >= nums ? "pagination-aft-disabled" : "pagination-aft"
//             }
//           >
//             前进
//           </div>
//         );
//         tpl.push(
//           <div>
//             跳至 <input type="text" />页
//           </div>
//         );
//         return tpl;
//       };
//       return () => {
//         return <div class="pagination-box">{templ()}</div>;
//       };
//     },
//   });
