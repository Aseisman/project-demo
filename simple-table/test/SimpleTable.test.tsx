import { mount, shallowMount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";
import { type ColumnsProps } from "../src/components/types";
import SimpleTable from "../src/components/SimpleTable";
import { timeFormat } from "../src/utils";
import Header from "../src/components/Headers";
import Body from "../src/components/Body";
import Line from "../src/components/Line";
import Pagination from "../src/components/Pagination/Pagination";
import Loading from "../src/components/Loading/Loading";

// 列表数据
const res = {
  data: [
    {
      name: "测试1",
      year: new Date().toString(),
      price: 567,
      nums: 32,
    },
    {
      name: "测试2",
      year: new Date().toString(),
      price: 334,
      nums: 12,
    },
    {
      name: "测试3",
      year: new Date().toString(),
      price: 755,
      nums: 34,
    },
    {
      name: "测试4",
      year: new Date().toString(),
      price: 234,
      nums: 77,
    },
    {
      name: "测试5",
      year: new Date().toString(),
      price: 123,
      nums: 1,
    },
    {
      name: "测试6",
      year: new Date().toString(),
      price: 200,
      nums: 21,
    },
    {
      name: "测试7",
      year: new Date().toString(),
      price: 200,
      nums: 11,
    },
    {
      name: "测试8",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
    {
      name: "测试9",
      year: new Date().toString(),
      price: 198,
      nums: 20,
    },
    {
      name: "测试10",
      year: new Date().toString(),
      price: 197,
      nums: 20,
    },
    {
      name: "测试11",
      year: new Date().toString(),
      price: 134,
      nums: 20,
    },
    {
      name: "测试12",
      year: new Date().toString(),
      price: 144,
      nums: 20,
    },
    {
      name: "测试13",
      year: new Date().toString(),
      price: 155,
      nums: 20,
    },
    {
      name: "测试14",
      year: new Date().toString(),
      price: 166,
      nums: 20,
    },
    {
      name: "测试15",
      year: new Date().toString(),
      price: 177,
      nums: 20,
    },
    {
      name: "测试16",
      year: new Date().toString(),
      price: 188,
      nums: 20,
    },
    {
      name: "测试17",
      year: new Date().toString(),
      price: 199,
      nums: 20,
    },
    {
      name: "测试18",
      year: new Date().toString(),
      price: 211,
      nums: 20,
    },
    {
      name: "测试19",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
    {
      name: "测试20",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
    {
      name: "测试21",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
    {
      name: "测试12",
      year: new Date().toString(),
      price: 144,
      nums: 20,
    },
    {
      name: "测试13",
      year: new Date().toString(),
      price: 155,
      nums: 20,
    },
    {
      name: "测试14",
      year: new Date().toString(),
      price: 166,
      nums: 20,
    },
    {
      name: "测试15",
      year: new Date().toString(),
      price: 177,
      nums: 20,
    },
    {
      name: "测试16",
      year: new Date().toString(),
      price: 188,
      nums: 20,
    },
    {
      name: "测试17",
      year: new Date().toString(),
      price: 199,
      nums: 20,
    },
    {
      name: "测试18",
      year: new Date().toString(),
      price: 211,
      nums: 20,
    },
    {
      name: "测试19",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
    {
      name: "测试20",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
    {
      name: "测试21",
      year: new Date().toString(),
      price: 200,
      nums: 20,
    },
  ],
  total: 21,
};

const getColumns=(type=1)=>{
  switch (type) {
    case 1:
      return  [
        {
          title: "名字",
          field: "name",
          key: "name",
          width: 100,
        },
        {
          title: "价格",
          field: "price",
          key: "price",
          width: 200,
        },
        {
          title: "数量",
          field: "nums",
          key: "nums",
          width: 300,
        },
        {
          title: "年份",
          field: "year",
          key: "year",
        },
      ];

    case 2:
      return [
        {
          title: "名字",
          field: "name",
          key: "name",
          width: 100,
        },
        {
          title: "价格",
          field: "price",
          key: "price",
          width: 200,
          sortOrder: "ascend",
        },
        {
          title: "数量",
          field: "nums",
          key: "nums",
          width: 300,
          render: (text: any, record: any, index: number) => {
            return <div style="background:grey;color:white;">{text}</div>;
          },
        },
        {
          title: "年份",
          field: "year",
          key: "year",
          formatter: (value: Date) => {
            let time = new Date(value).getDate();
            return timeFormat("YYYY-mm-dd HH:MM", time);
          },
        },
      ];

    default:
      return []
  }
  
}

const wait = (time) => {
  return new Promise((resolve) => setTimeout(resolve, time));
};

describe("SimpleTable", () => {
  // columns
  test("check orderSort as 'ascend' in some item", async () => {
    const columns: ColumnsProps[] = getColumns();
    columns[1].sortOrder='ascend';

    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    const items = wrapper.findComponent({ name: "Body" });
    const data = items.props().data;
    let flag = true;
    data.reduce((a, c) => {
      if (a.price > c.price) {
        flag = false;
      }
      return c;
    }, 0);
    expect(flag).toBe(true);
  });
  test("check orderSort as 'descend' in some item", async () => {
    const columns: ColumnsProps[] = getColumns();
    columns[1].sortOrder='descend';

    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    const items = wrapper.findComponent({ name: "Body" });
    const data = items.props().data;
    let flag = true;
    data.reduce((a, c) => {
      if (a.price < c.price) {
        flag = false;
      }
      return c;
    }, 0);
    expect(flag).toBe(true);
  });
  test("check orderSort as undefined in some item", async () => {
    const columns: ColumnsProps[] = getColumns();
    columns[1].sortOrder=undefined;
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    const items = wrapper.findComponent({ name: "Body" });
    const data = items.props().data;
    let flag = true;
    data.map((item,index)=>{
      if(item.price!=res.data[index].price){
        flag = false;
      }
    })
    expect(flag).toBe(true);
  });
  test("sort by sb", async () => {
    const columns: ColumnsProps[] = getColumns();
    columns[1].sorter=(a, b) => a - b;
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    const priceIcon = wrapper.find(".sort-icon");

    priceIcon.trigger("click");

    await wait(500);

    let items = wrapper.findComponent({ name: "Body" });
    let data = items.props().data;
    let flag = true;
    data.reduce((a, c) => {
      if (a.price > c.price) {
        flag = false;
      }
      return c;
    }, 0);
    expect(flag).toBe(true);

    priceIcon.trigger("click");

    await wait(500);
    items = wrapper.findComponent({ name: "Body" });
    data = items.props().data;
    flag = true;
    data.reduce((a, c) => {
      if (a.price < c.price) {
        flag = false;
      }
      return c;
    }, 0);
    expect(flag).toBe(true);

    priceIcon.trigger("click");
    await wait(500);
    items = wrapper.findComponent({ name: "Body" });
    data = items.props().data;
    flag = true;
    data.map((item,index)=>{
      if(item.price!=res.data[index].price){
        flag = false;
      }
    })
    expect(flag).toBe(true);

  });
  test("multiSort and sort by sb", async () => {
    const columns: ColumnsProps[] = getColumns();
    columns[1].sorter=(a, b) => a - b;
    columns[2].sorter=(a, b) => a - b;
    
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    const HeaderWrapper = wrapper.findComponent(Header).findAll(".sort-icon");

    HeaderWrapper[0].trigger("click");
    await wait(500);
    let items = wrapper.findComponent({ name: "Body" });
    let data = items.props().data;
    let flag = true;
    data.reduce((a, c) => {
      if (a.price > c.price) {
        flag = false;
      }
      return c;
    }, 0);
    expect(flag).toBe(true);

    HeaderWrapper[1].trigger("click");
    await wait(500);
    items = wrapper.findComponent({ name: "Body" });
    data = items.props().data;
    flag = true;
    data.reduce((a, c) => {
      if (a.nums > c.nums) {
        flag = false;
      }
      return c;
    }, 0);
    expect(flag).toBe(true);
  });
  test("check render, formatter and width in some item", () => {
    const columns: ColumnsProps[] = getColumns();
    columns[1].sortOrder='ascend';
    columns[2].render=(text: any, record: any, index: number) => {
      return <div class='test-cls' style="background:grey;color:white;">{text}</div>;
    };
    columns[3].formatter=(value: Date) => {
      let time = new Date(value).getDate();
      return timeFormat("YYYY-mm-dd HH:MM", time);
    };
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    expect(wrapper.findAll('.test-cls').length!==0).toBe(true);

    const items = wrapper.findComponent({ name: "Header" });
    let flag = true;
    items.props().columns.forEach((column, index) => {
      if (column.width != columns[index].width) {
        flag = false;
      }
    });
    expect(flag).toBe(true);
  });
  // pagination
  test("check pagination in table", async () => {
    const columns: ColumnsProps[] = getColumns(2);
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
        pagination: {
          total: 10,
          pageSize: 4,
        },
      },
    });
    let items = wrapper.findComponent({ name: "Body" });
    let data = items.props().data;
    expect(data.length==4).toBe(true);
    await wrapper.setProps({
      pagination: {
        total: 10,
        pageSize: 5,
      },
    });

    items = wrapper.findComponent({ name: "Body" });
    data = items.props().data;
    expect(data.length==5).toBe(true);

    
    await wrapper.setProps({
      pagination: {
        total: 10,
        current: 1,
        pageSize: 5,
      },
    });

    items = wrapper.findComponent({ name: "Pagination" });
    data = items.props().pagination.current;
    expect(data==1).toBe(true);

    await wrapper.setProps({
      pagination: {
        total: 10,
        current: 2,
        pageSize: 5,
      },
    });

    items = wrapper.findComponent({ name: "Pagination" });
    data = items.props().pagination.current;
    expect(data==2).toBe(true);


    const preButton = wrapper.find("#pre-btn");
    preButton.trigger("click");

    items = wrapper.findComponent({ name: "Pagination" });
    data = items.props().pagination.current;
    expect(data==2).toBe(true);


    const nextButton = wrapper.find("#next-btn");
    nextButton.trigger("click");

    items = wrapper.findComponent({ name: "Pagination" });
    data = items.props().pagination.current;
    expect(data==2).toBe(true);

  });
  test("check pagination as false in table", async () => {
    const columns: ColumnsProps[] = getColumns(2);
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
        pagination: false,
      },
    });
    let items = wrapper.findComponent({ name: "Body" });
    let data = items.props().data;
    expect(data.length==res.data.length).toBe(true);
  });
  test("set num in paginations' input", async () => {
    const columns: ColumnsProps[] = getColumns(2);
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
        pagination: {
          total: 10,
          pageSize: 4,
        },
      },
    });
    const inputDom = wrapper.find("input");

    await inputDom.setValue(2);
    await inputDom.setValue(1);
    await wait(1000);

    expect(inputDom.element.value).toBe("1");

    await inputDom.setValue("");
    await wait(1000);

    expect(wrapper.findComponent(Pagination).vm.pagination.current).toBe(0);

    await inputDom.setValue("-1");
    await wait(1000);

    expect(wrapper).toThrowErrorMatchingSnapshot();
  });

  // body
  test("nodata in body", () => {
    const columns: ColumnsProps[] = getColumns();
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
      },
    });
    let items = wrapper.findComponent({ name: "Body" });
    let data = items.props().data;
    expect(data.length==0).toBe(true);
  });
  // onchange
  test("emit onchange for simpleTable", () => {
    const columns: ColumnsProps[] = getColumns();
    const onChange = vi.fn();
    const wrapper = mount(SimpleTable, {
      props: {
        columns,
        data: res.data,
        pagination: {
          total: 10,
          pageSize: 4,
        },
        onChange: (...arg) => {
          console.log(arg);
          console.log("onchange");
          onChange();
        },
      },
    });
    const preButton = wrapper.find("#next-btn");
    preButton.trigger("click");
    expect(onChange).toHaveBeenCalled();
  });
  // mount
  test("mount component", () => {
    const wrapper = shallowMount(SimpleTable, {
      props: {},
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
  test("mount component with columns", () => {
    const columns: ColumnsProps[] = getColumns();
    const wrapper = shallowMount(SimpleTable, {
      props: {
        columns,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("mount component with data", () => {
    const columns: ColumnsProps[] = getColumns();
    const wrapper = shallowMount(SimpleTable, {
      props: {
        columns,
        data: res.data,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("mount Loading", async () => {
    const columns: ColumnsProps[] = getColumns();
    const wrapper = shallowMount(SimpleTable, {
      props: {
        columns,
        data: res.data,
        loading: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    const LoadingWrapper = shallowMount(Loading, {
      props: {
        loading: true,
      },
    });
    expect(LoadingWrapper.html()).toMatchSnapshot();
  });
});
