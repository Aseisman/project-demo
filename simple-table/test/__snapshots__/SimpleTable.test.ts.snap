// Vitest Snapshot v1

exports[`mount component 1`] = `
"<table style=\\"\\">
  <loading-stub loading=\\"false\\"></loading-stub>
</table>"
`;
